File: csharp/src/BrValidator/Ie/Ba.cs
using System.Collections.Generic;
using System.Text.RegularExpressions;
using BrValidator.Internal;

namespace BrValidator.Ie
{
    /// <summary>
    /// BA: verified against the official SEFAZ-BA "Roteiro de Crítica da
    /// Inscrição Estadual" (sintegra.gov.br/Cad_Estados/cad_BA.html),
    /// including its four worked examples (123456-63, 612345-57,
    /// 1000003-06 mod-10/mod-11 x 8/9-digit variants).
    ///
    /// Bahia has two lengths (8 or 9 digits) and, within each, two moduli:
    /// módulo 10 when the discriminating digit (the 1st digit for 8-digit
    /// IEs, the 2nd for 9-digit IEs) is one of 0,1,2,3,4,5,8, and módulo 11
    /// when it is 6, 7 or 9. The last check digit is calculated first (from
    /// the base digits alone), then the first check digit is calculated
    /// from the base plus that digit.
    /// </summary>
    internal static class Ba
    {
        private static readonly Regex AllowedChars = new Regex(@"^[\d\-\s]+$", RegexOptions.Compiled);
        private static readonly HashSet<char> Mod10Digits = new HashSet<char> { '0', '1', '2', '3', '4', '5', '8' };
        private static readonly HashSet<char> Mod11Digits = new HashSet<char> { '6', '7', '9' };

        public static string Normalize(string value) => Shared.RemoveNonDigits(value);

        private static int CheckDigitFor(int sum, bool useMod11) =>
            useMod11 ? Mod11.Mod11CheckDigit(sum) : Mod11.Mod10CheckDigit(sum);

        private static int[] DescendingWeights(int length)
        {
            int[] weights = new int[length];
            for (int i = 0; i < length; i++)
            {
                weights[i] = length + 1 - i;
            }
            return weights;
        }

        public static bool IsValid(string value)
        {
            if (!AllowedChars.IsMatch(value))
            {
                return false;
            }

            string v = Normalize(value);

            if (v.Length != 8 && v.Length != 9)
            {
                return false;
            }

            char discriminant = v.Length == 8 ? v[0] : v[1];
            bool useMod11 = Mod11Digits.Contains(discriminant);

            if (!useMod11 && !Mod10Digits.Contains(discriminant))
            {
                return false;
            }

            int baseLength = v.Length - 2;
            string baseDigits = v.Substring(0, baseLength);
            int lastDigit = CheckDigitFor(Mod11.WeightedSum(baseDigits, DescendingWeights(baseLength)), useMod11);

            if (lastDigit != (v[v.Length - 1] - '0'))
            {
                return false;
            }

            string baseWithLastDigit = baseDigits + lastDigit;
            int firstDigit = CheckDigitFor(
                Mod11.WeightedSum(baseWithLastDigit, DescendingWeights(baseLength + 1)), useMod11);

            return firstDigit == (v[v.Length - 2] - '0');
        }

        public static string Format(string value)
        {
            string v = Normalize(value);

            if (v.Length != 8 && v.Length != 9)
            {
                return value;
            }

            return $"{v.Substring(0, v.Length - 2)}-{v.Substring(v.Length - 2, 2)}";
        }
    }
}

File: src/ie/states/ba.ts
import { removeNonDigits } from "../../shared/normalize.js";
import { weightedSum, mod10CheckDigit, mod11CheckDigit } from "../../shared/mod11.js";

// Verified against the official SEFAZ-BA "Roteiro de Crítica da Inscrição
// Estadual" (sintegra.gov.br/Cad_Estados/cad_BA.html), including its four
// worked examples (123456-63, 612345-57, 1000003-06 mod-10/mod-11 x
// 8/9-digit variants).
//
// Bahia has two lengths (8 or 9 digits) and, within each, two moduli: módulo
// 10 when the discriminating digit (the 1st digit for 8-digit IEs, the 2nd
// for 9-digit IEs) is one of 0,1,2,3,4,5,8, and módulo 11 when it is 6, 7 or
// 9. The last check digit is calculated first (from the base digits alone),
// then the first check digit is calculated from the base plus that digit.
const MOD10_DIGITS = new Set(["0", "1", "2", "3", "4", "5", "8"]);
const MOD11_DIGITS = new Set(["6", "7", "9"]);

function normalize(value: string): string {
  return removeNonDigits(value);
}

function checkDigitFor(sum: number, useMod11: boolean): number {
  return useMod11 ? mod11CheckDigit(sum) : mod10CheckDigit(sum);
}

function isValid(value: string): boolean {
  if (!/^[\d\-\s]+$/.test(value)) {
    return false;
  }

  const ie = normalize(value);

  if (ie.length !== 8 && ie.length !== 9) {
    return false;
  }

  const discriminant = ie.length === 8 ? ie[0] : ie[1];

  if (!discriminant) {
    return false;
  }

  const useMod11 = MOD11_DIGITS.has(discriminant);

  if (!useMod11 && !MOD10_DIGITS.has(discriminant)) {
    return false;
  }

  const baseLength = ie.length - 2;
  const base = ie.slice(0, baseLength);
  const lastDigitWeights = Array.from({ length: baseLength }, (_, i) => baseLength + 1 - i);
  const lastDigit = checkDigitFor(weightedSum(base, lastDigitWeights), useMod11);

  if (lastDigit !== Number(ie[ie.length - 1])) {
    return false;
  }

  const firstDigitWeights = Array.from({ length: baseLength + 1 }, (_, i) => baseLength + 2 - i);
  const firstDigit = checkDigitFor(weightedSum(base + lastDigit, firstDigitWeights), useMod11);

  return firstDigit === Number(ie[ie.length - 2]);
}

function format(value: string): string {
  const ie = normalize(value);

  if (ie.length !== 8 && ie.length !== 9) {
    return value;
  }

  return `${ie.slice(0, -2)}-${ie.slice(-2)}`;
}

export const BA = {
  isValid,
  normalize,
  format,
};
